### PR TITLE
[time] Eliminate libc timezone bloat (~9.5KB flash ESP32, ~2% RAM on ESP8266)

### PR DESCRIPTION
**Pull request by bdraco**
# What does this implement/fix?

This PR replaces libc's timezone handling with a lightweight custom implementation, recovering ~9.5KB flash on ESP32 and ~2% of usable RAM on ESP8266. It is mostly tests (126 of them) because time handling needs to be rock solid and consistent across all platforms. The majority of those tests cover the C++ POSIX TZ string parser, which is **temporary bridge code** — after #14233 merges, the parser is solely used to handle timezone strings from Home Assistant clients older than 2026.3.0 that haven't been updated to send the pre-parsed `ParsedTimezone` protobuf struct. The parser and its tests will be removed before ESPHome 2026.9.0. The remaining tests cover the time conversion functions (`epoch_to_local_tm`, `is_in_dst`, `strptime`) which are permanent.

**This is step 1 of a long-term plan to eliminate all timezone parsing from device firmware entirely.** See [esphome/backlog#91](https://github.com/esphome/backlog/issues/91) for the full roadmap. Since all timezone data originates from Python (codegen or aioesphomeapi), future steps will construct the `ParsedTimezone` struct in Python and send it pre-computed — removing the parser from the binary completely. This PR establishes the target struct format and the conversion functions that will remain.

> **Note:** The `posix_tz.cpp` parser code is **temporary bridge code** that exists solely for backward compatibility. Now that [aioesphomeapi#1505](https://github.com/esphome/aioesphomeapi/pull/1505) has merged, once #14233 merges the only remaining use of the C++ parser is to handle the string-based `timezone` field from older Home Assistant clients that haven't been updated to send the pre-parsed `ParsedTimezone` struct. After a 6-month transition period (target removal: **ESPHome 2026.9.0**), the parser will be removed entirely. The conversion functions (`epoch_to_local_tm`, `is_in_dst`, etc.) and the `ParsedTimezone`/`DSTRule` structs will remain permanently.

## Summary

We were using libc's `localtime()`, `mktime()`, and `tzset()` to handle timezones, but these functions pulled in massive baggage that **only libc itself needed**:

- The entire `scanf` family (~9.8KB flash) - just to parse the TZ environment variable
- Environment variable infrastructure (`setenv`/`getenv`) - only used by libc's `tzset()`
- Double storage of timezone data - once in libc's BSS via the TZ env var, again as a `std::string` in our component

**Nothing else in the codebase used any of this.** On embedded targets, no ESPHome code uses environment variables—only the host platform even has them. It was pure overhead on embedded targets.

We already had `ESPTime` functions to abstract all this away—except they delegated to libc internally, which made them inconsistent across platforms (different newlib versions on ESP32/ESP8266/RP2040/LibreTiny, plus glibc/musl on host). Now they're consistent everywhere.

This PR parses POSIX TZ strings directly, storing the result in a compact 32-byte struct. The libc functions are no longer called, so the linker drops all that dead weight.

### Why Embedded Systems Don't Need TZ Environment Variables

Environment variables don't make sense on embedded systems—there's no subprocess to pass them to. Additionally, there are [reports of memory leaks](https://www.reddit.com/r/Esphome/comments/1auj7ia/handling_utc_time_in_esphome_lambda/) in `setenv`/`tzset` on both Arduino and ESP-IDF frameworks that can crash devices:
- **No shell** - devices boot straight into firmware
- **No other processes** - ESPHome is the only thing running
- **No persistent environment** - `setenv()` only exists in RAM until reboot
- **No external consumers** - nothing else calls `getenv("TZ")`

The old code path was: ESPHome sets TZ env var → libc reads TZ → libc parses TZ with scanf → libc returns local time

But ESPHome already knew the timezone—it just set it! The entire env var dance was just a way to pass data to libc so libc could parse it with `scanf`. Now we skip the middleman.

### Memory Savings

| Platform   | Flash Savings | RAM Savings |
|------------|---------------|-------------|
| ESP32-IDF  | ~9.5KB        | 136 bytes   |
| ESP8266    | ~4.1KB        | ~850 bytes (272 static + 578 heap) |
| LibreTiny  | ~4.7KB        | 148 bytes   |
| RP2040     | ~3.7KB        | 148 bytes   |

For context on ESP32, linked FreeRTOS code is ~16KB. The timezone infrastructure we're removing (~9.5KB) is nearly **2/3 the size of the entire used RTOS**—just to parse a string that ESPHome already knew.

On ESP8266 with ~45KB usable RAM after SDK, **850 bytes is nearly 2%** - recovered by not linking unused functions, not storing env vars nothing will ever read, and not double-storing the timezone.

### Configs Without Timezone Also Benefit

Even configs that don't use a timezone save **~656 bytes flash** on ESP8266. Previously, the time component called libc's `localtime()` even without a timezone configured, which just returned UTC anyway but still linked all the libc bloat. Now those calls are gone.

### Bonus: Consistent Cross-Platform Behavior

Different libc implementations (newlib, glibc, musl) handle timezone edge cases slightly differently. By parsing TZ strings ourselves with 126 unit tests, timezone behavior is now **identical** across ESP32, ESP8266, RP2040, LibreTiny, and Host platforms.

### Logging Format Change

The `dump_config()` output now shows human-readable timezone information:

**Before:**
```
[C][time:027]: Timezone: 'CST6CDT,M3.2.0,M11.1.0'
```

**After:**
```
[C][time:030]: Timezone: UTC-6:00 (DST UTC-5:00)
[C][time:030]: Timezone: UTC+5:30
```

### Long-term Roadmap

This PR is **step 1** of [esphome/backlog#91](https://github.com/esphome/backlog/issues/91):

1. **Step 1 (this PR):** Replace libc's bloated timezone parser with a lightweight custom parser → establishes the `ParsedTimezone` struct and conversion functions
2. **Step 2 ✅ (this PR):** Parse timezone in Python codegen → emit `ParsedTimezone` struct directly in generated C++, no runtime parsing needed for configured timezones
3. **Step 3 ✅ (#14233 + [aioesphomeapi#1505](https://github.com/esphome/aioesphomeapi/pull/1505)):** aioesphomeapi sends pre-parsed `ParsedTimezone` fields over protobuf → no runtime parsing for HA-provided timezones either
4. **Step 4 (#14237, ESPHome 2026.9.0):** Remove `parse_posix_tz()` and all parsing helpers from firmware → device only contains conversion functions (`epoch_to_local_tm`, `is_in_dst`). The C++ parser (`posix_tz.cpp`) is only needed during the transition period while older Home Assistant clients still send the timezone as a string instead of the pre-parsed struct. Draft PR #14237 is open to verify flash/RAM savings.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected) — [policy](https://developers.esphome.io/contributing/code/#what-constitutes-a-c-breaking-change)
- [ ] Developer breaking change (an API change that could break external components) — [policy](https://developers.esphome.io/contributing/code/#what-is-considered-public-c-api)
- [x] Undocumented C++ API change (removal or change of undocumented public methods that lambda users may depend on) — [policy](https://developers.esphome.io/contributing/code/#c-user-expectations)
- [x] Code quality improvements to existing code or addition of tests
- [ ] Other

**Breaking change (extremely limited scope):** We override libc's `localtime()` and `localtime_r()` on embedded platforms, so most user lambdas work unchanged. However, lambdas using `::mktime()`, bare `::strftime()`, or calling `setenv("TZ", ...)`/`tzset()` directly would be affected.

These edge cases are unlikely—all docs examples use ESPHome's time API, and anyone calling `setenv`/`tzset` directly is likely hitting the memory leak in not-well-tested code paths on ESP platforms anyway.

**Docs update:** The [modbus_controller docs](https://esphome.io/components/modbus_controller.html) have an EPEVER example that uses `::localtime()` in a lambda. Updated in esphome/esphome-docs#6011 to use ESPHome's `ESPTime` API (recommended best practice, though the old code will still work).

**Verified:** All external libraries in platformio.ini were checked - none use `localtime()`, `mktime()`, `tzset()`, or `strftime()`:
- qr-code-generator-library, arduino-MLX90393, HaierProtocol, dsmr_parser, Crypto-no-arduino
- lvgl, noise-c, Improv, pngle, AsyncMqttClient-esphome
- FastLED, TM1651, MideaUART, HeatpumpIR, ArduinoJson
- ESPAsyncWebServer, NeoPixelBus, esp_wireguard, ESP32-audioI2S, ESPMicroSpeechFeatures

**Related issue or feature (if applicable):**

- Significant flash/RAM savings for all devices using the time component
- Eliminates platform-specific timezone behavior differences
- Part of [esphome/backlog#91](https://github.com/esphome/backlog/issues/91)

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#6011 (updates EPEVER lambda example to use ESPTime API)

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [x] ESP8266
- [x] RP2040
- [x] BK72xx
- [x] RTL87xx
- [x] LN882x
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
# No configuration changes required - existing timezone configs work unchanged
time:
  - platform: homeassistant
    id: ha_time
    timezone: "EST5EDT,M3.2.0,M11.1.0"
```

### Test button to verify `localtime()` override works:

```yaml
time:
  - platform: homeassistant
    timezone: "CST6CDT,M3.2.0,M11.1.0"

button:
  - platform: template
    name: "Test localtime Override"
    on_press:
      - lambda: |-
          // Test that ::localtime() returns local time (not UTC) via our override
          time_t now = ::time(nullptr);
          struct tm *time_info = ::localtime(&now);
          ESP_LOGI("test", "::localtime(): %04d-%02d-%02d %02d:%02d:%02d",
                   time_info->tm_year + 1900, time_info->tm_mon + 1, time_info->tm_mday,
                   time_info->tm_hour, time_info->tm_min, time_info->tm_sec);
          // Compare with ESPTime for verification
          auto t = ESPTime::from_epoch_local(now);
          ESP_LOGI("test", "ESPTime:        %04d-%02d-%02d %02d:%02d:%02d",
                   t.year, t.month, t.day_of_month, t.hour, t.minute, t.second);
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/time/real_time_clock.cpp [hunk 1]
```diff
@@ -89,9 +94,13 @@ void RealTimeClock::synchronize_epoch_(uint32_t epoch) {
...
-  setenv("TZ", this->timezone_.c_str(), 1);
-  tzset();
+void RealTimeClock::apply_timezone_(const char *tz) {
+  // Parse the POSIX TZ string using our custom parser to avoid pulling in scanf (~7.6KB)
+  if (!parse_posix_tz(tz, this->parsed_tz_)) {
+    ESP_LOGW(TAG, "Failed to parse timezone: %s", tz);
+    // Reset to UTC on parse failure
+    this->parsed_tz_ = ParsedTimezone{};
+  }
 }
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/time/real_time_clock.cpp:138]: nothing actually uses that

File: esphome/components/time/posix_tz.cpp [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+
+    // Second rule is required per POSIX
+    if (*p != ',') {
+      return false;
+    }
+    p++;
+    if (!internal::parse_dst_rule(p, result.dst_end)) {
+      return false;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/time/posix_tz.cpp]: 300eea034b

File: tests/components/time/posix_tz_parser.cpp [hunk 3]
```diff
@@ -0,0 +1,977 @@
...
+  EXPECT_EQ(tz.dst_offset_seconds, 4 * 3600);  // Default: std - 1 hour
+}
+
+TEST(PosixTzParser, TrailingCharactersIgnored) {
+  ParsedTimezone tz;
+  // Trailing characters after valid TZ should be ignored (parser stops at end of valid input)
+  // This matches libc behavior
+  ASSERT_TRUE(parse_posix_tz("EST5", tz));
+  EXPECT_EQ(tz.std_offset_seconds, 5 * 3600);
+}
```
[the remark this answers is not in the document]
Comment (bdraco) [reply tests/components/time/posix_tz_parser.cpp:426]: 300eea034b

File: esphome/components/time/posix_tz.cpp [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+    return true;  // No DST
+  }
+
+  // If next char is comma, there's no DST name but there are rules (invalid)
+  if (*p == ',') {
+    return false;
+  }
+
+  if (!internal::skip_tz_name(p)) {
+    return true;  // No valid DST name, no DST
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/time/posix_tz.cpp]: 300eea034b

File: esphome/components/time/posix_tz.cpp [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+    p++;
+    minutes = parse_uint(p);
+    if (*p == ':') {
+      p++;
+      seconds = parse_uint(p);
+    }
+  }
+
+  return sign * (hours * 3600 + minutes * 60 + seconds);
+}
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/time/posix_tz.cpp:169]: probably not a case that ever comes up

File: esphome/components/time/real_time_clock.cpp [hunk 6]
```diff
@@ -25,7 +25,17 @@ RealTimeClock::RealTimeClock() = default;
...
+  int std_hours = -this->parsed_tz_.std_offset_seconds / 3600;
+  int std_mins = abs(this->parsed_tz_.std_offset_seconds % 3600) / 60;
+  ESP_LOGCONFIG(TAG, "Timezone: UTC%+d:%02d", std_hours, std_mins);
+  if (this->parsed_tz_.has_dst) {
+    int dst_hours = -this->parsed_tz_.dst_offset_seconds / 3600;
+    ESP_LOGCONFIG(TAG, "  DST: UTC%+d, M%d.%d.%d/%" PRId32 " - M%d.%d.%d/%" PRId32, dst_hours,
+                  this->parsed_tz_.dst_start.month, this->parsed_tz_.dst_start.week,
+                  this->parsed_tz_.dst_start.day_of_week, this->parsed_tz_.dst_start.time_seconds / 3600,
+                  this->parsed_tz_.dst_end.month, this->parsed_tz_.dst_end.week, this->parsed_tz_.dst_end.day_of_week,
+                  this->parsed_tz_.dst_end.time_seconds / 3600);
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/time/real_time_clock.cpp]: 300eea034b

File: esphome/components/time/real_time_clock.cpp [hunk 7]
```diff
@@ -89,9 +94,13 @@ void RealTimeClock::synchronize_epoch_(uint32_t epoch) {
 }
 
 #ifdef USE_TIME_TIMEZONE
-void RealTimeClock::apply_timezone_() {
-  setenv("TZ", this->timezone_.c_str(), 1);
-  tzset();
+void RealTimeClock::apply_timezone_(const char *tz) {
+  // Parse the POSIX TZ string using our custom parser to avoid pulling in scanf (~7.6KB)
+  if (!parse_posix_tz(tz, this->parsed_tz_)) {
+    ESP_LOGW(TAG, "Failed to parse timezone: %s", tz);
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/time/real_time_clock.cpp:132]: 300eea034b

File: tests/components/time/posix_tz_parser.cpp [hunk 8]
```diff
@@ -0,0 +1,977 @@
...
+
+  // Our implementation
+  struct tm our_tm;
+  epoch_to_local_tm(epoch, tz, &our_tm);
+
+  // libc implementation
+  setenv("TZ", tz_str, 1);
+  tzset();
+  struct tm *libc_tm = localtime(&epoch);
+
```
[the remark this answers is not in the document]
Comment (bdraco) [reply tests/components/time/posix_tz_parser.cpp:806]: 300eea034b

File: esphome/components/time/real_time_clock.h [hunk 9]
```diff
@@ -20,26 +23,31 @@ class RealTimeClock : public PollingComponent {
...
-  /// Set the time zone from raw buffer, only if it differs from the current one.
-  void set_timezone(const char *tz, size_t len) {
-    if (this->timezone_.length() != len || memcmp(this->timezone_.c_str(), tz, len) != 0) {
-      this->timezone_.assign(tz, len);
-      this->apply_timezone_();
-    }
-  }
+  /// Set the time zone from a null-terminated string with known length.
+  /// The length parameter is ignored since our parser uses null-terminated strings.
+  void set_timezone(const char *tz, size_t /*len*/) { this->apply_timezone_(tz); }
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/time/real_time_clock.h]: we don't want to allocate a string, bad idea...

[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/time/real_time_clock.h]: used a stack buf

File: esphome/components/time/real_time_clock.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/time/real_time_clock.cpp]: I took it back out aioesphomeapi and esphome always use M format. this was overkill. The parser supports all the formats for compat but realistically we just don't have them.

File: esphome/components/time/posix_tz.cpp [hunk 10]
```diff
@@ -0,0 +1,443 @@
+#include "posix_tz.h"
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/time/posix_tz.cpp:5]: removed format_dst_rule

File: esphome/components/time/posix_tz.cpp [hunk 11]
```diff
@@ -0,0 +1,443 @@
...
+
+  // Parse standard offset (required)
+  if (!*p || (!std::isdigit(static_cast<unsigned char>(*p)) && *p != '+' && *p != '-')) {
+    return false;
+  }
+  result.std_offset_seconds = internal::parse_offset(p);
+
+  // Check for DST name
+  if (!*p) {
+    return true;  // No DST
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/time/posix_tz.cpp:400]: removed format_dst_rule

File: esphome/components/time/real_time_clock.h [hunk 12]
```diff
@@ -20,26 +23,39 @@ class RealTimeClock : public PollingComponent {
...
-      this->timezone_.assign(tz, len);
-      this->apply_timezone_();
-    }
+    // Stack buffer - TZ strings are typically short but allow up to 128
+    char buf[128];
+    if (len >= sizeof(buf))
+      len = sizeof(buf) - 1;
+    memcpy(buf, tz, len);
+    buf[len] = '\0';
+    this->apply_timezone_(buf);
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/time/real_time_clock.h:42]: added

File: tests/components/time/posix_tz_parser.cpp [hunk 13]
```diff
@@ -0,0 +1,999 @@
...
+  ASSERT_TRUE(parse_posix_tz(tz_str, tz));
+
+  // Our implementation
+  struct tm our_tm;
+  epoch_to_local_tm(epoch, tz, &our_tm);
+
+  // libc implementation
+  setenv("TZ", tz_str, 1);
+  tzset();
+  struct tm *libc_tm = localtime(&epoch);
```
[the remark this answers is not in the document]
Comment (bdraco) [reply tests/components/time/posix_tz_parser.cpp]: fixed

File: esphome/components/time/posix_tz.cpp [hunk 14]
```diff
@@ -0,0 +1,443 @@
...
+
+namespace internal {
+
+// Helper to parse an unsigned integer from string, updating pointer
+static uint32_t parse_uint(const char *&p) {
+  uint32_t value = 0;
+  while (std::isdigit(static_cast<unsigned char>(*p))) {
+    value = value * 10 + (*p - '0');
+    p++;
+  }
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/time/posix_tz.cpp:27]: not a case that actually comes up

File: esphome/components/time/real_time_clock.cpp [hunk 15]
```diff
@@ -25,7 +25,18 @@ RealTimeClock::RealTimeClock() = default;
 
 void RealTimeClock::dump_config() {
 #ifdef USE_TIME_TIMEZONE
-  ESP_LOGCONFIG(TAG, "Timezone: '%s'", this->timezone_.c_str());
+  int std_hours = -this->parsed_tz_.std_offset_seconds / 3600;
+  int std_mins = abs(this->parsed_tz_.std_offset_seconds % 3600) / 60;
+  ESP_LOGCONFIG(TAG, "Timezone: UTC%+d:%02d", std_hours, std_mins);
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/time/real_time_clock.cpp]: de06b36544

File: esphome/components/time/real_time_clock.cpp [hunk 16]
```diff
@@ -89,9 +95,20 @@ void RealTimeClock::synchronize_epoch_(uint32_t epoch) {
...
+  if (tz == nullptr) {
+    ESP_LOGW(TAG, "Failed to parse timezone: (null)");
+    this->parsed_tz_ = ParsedTimezone{};
+    return;
+  }
+
+  // Parse the POSIX TZ string using our custom parser
+  if (!parse_posix_tz(tz, this->parsed_tz_)) {
+    ESP_LOGW(TAG, "Failed to parse timezone: %s", tz);
+    // Reset to UTC on parse failure
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/time/real_time_clock.cpp]: fixed from_epoch_local and timezone_offset

[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/time/real_time_clock.cpp]: 9e6e8a7ecb

File: esphome/components/time/real_time_clock.cpp [hunk 17]
```diff
@@ -25,7 +25,18 @@ RealTimeClock::RealTimeClock() = default;
...
 void RealTimeClock::dump_config() {
 #ifdef USE_TIME_TIMEZONE
-  ESP_LOGCONFIG(TAG, "Timezone: '%s'", this->timezone_.c_str());
+  int std_hours = -this->parsed_tz_.std_offset_seconds / 3600;
+  int std_mins = abs(this->parsed_tz_.std_offset_seconds % 3600) / 60;
+  ESP_LOGCONFIG(TAG, "Timezone: UTC%+d:%02d", std_hours, std_mins);
+  if (this->parsed_tz_.has_dst) {
+    int dst_hours = -this->parsed_tz_.dst_offset_seconds / 3600;
+    // Always use M format - tzdata and aioesphomeapi only generate M format rules
+    ESP_LOGCONFIG(TAG, "  DST: UTC%+d, M%d.%d.%d/%" PRId32 " - M%d.%d.%d/%" PRId32, dst_hours,
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/time/real_time_clock.cpp]: fixed

File: esphome/components/time/posix_tz.cpp [hunk 18]
```diff
@@ -0,0 +1,450 @@
...
+    days += days_in_month(year, m);
+  }
+  return days;
+}
+
+// Calculate days from epoch to Jan 1 of given year
+static int64_t __attribute__((noinline)) days_to_year_start(int year) {
+  int64_t days = 0;
+  for (int y = 1970; y < year; y++) {
+    days += days_in_year(y);
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/time/posix_tz.cpp:288]: not ever ahppening

File: esphome/core/time.cpp [hunk 19]
```diff
@@ -67,56 +66,129 @@ std::string ESPTime::strftime(const char *format) {
...
+      return false;
+
+    esp_time.year = v1;
+    esp_time.month = v2;
+    esp_time.day_of_month = v3;
+
+    if (p == end) {
+      // YYYY-MM-DD (date only)
+      return true;
+    }
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/core/time.cpp:125]: crazy.. never happens

File: esphome/components/time/posix_tz.cpp [hunk 20]
```diff
@@ -0,0 +1,452 @@
...
+#include <cctype>
+
+namespace esphome::time {
+
+// Global timezone for ESPTime::from_epoch_local() to use
+static ParsedTimezone global_tz_{};
+
+void set_global_tz(const ParsedTimezone &tz) { global_tz_ = tz; }
+
+const ParsedTimezone &get_global_tz() { return global_tz_; }
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/time/posix_tz.cpp:16]: huh? timezone doesn't change every few minutes.. thats nuts

File: esphome/core/time.cpp [hunk 21]
```diff
@@ -67,56 +66,129 @@ std::string ESPTime::strftime(const char *format) {
...
+      return false;
+
+    esp_time.hour = v1;
+    esp_time.minute = v2;
+
+    if (p == end) {
+      // HH:MM
+      esp_time.second = 0;
+      return true;
+    }
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/core/time.cpp:172]: wild edge case

File: esphome/components/time/real_time_clock.cpp [hunk 22]
```diff
@@ -25,7 +25,23 @@ RealTimeClock::RealTimeClock() = default;
...
+    int dst_mins = std::abs(tz.dst_offset_seconds % 3600) / 60;
+    // Transition times (when DST starts/ends)
+    int start_time_hours = tz.dst_start.time_seconds / 3600;
+    int start_time_mins = std::abs(tz.dst_start.time_seconds % 3600) / 60;
+    int end_time_hours = tz.dst_end.time_seconds / 3600;
+    int end_time_mins = std::abs(tz.dst_end.time_seconds % 3600) / 60;
+    // Always use M format - tzdata and aioesphomeapi only generate M format rules
+    ESP_LOGCONFIG(TAG, "  DST: UTC%+d:%02d, M%d.%d.%d/%d:%02d - M%d.%d.%d/%d:%02d", dst_hours, dst_mins,
+                  tz.dst_start.month, tz.dst_start.week, tz.dst_start.day_of_week, start_time_hours, start_time_mins,
+                  tz.dst_end.month, tz.dst_end.week, tz.dst_end.day_of_week, end_time_hours, end_time_mins);
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/time/real_time_clock.cpp]: never will happen

[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/time/real_time_clock.cpp]: never happens in practice

File: esphome/components/time/real_time_clock.h [hunk 23]
```diff
@@ -20,26 +23,43 @@ class RealTimeClock : public PollingComponent {
...
-      this->timezone_.assign(tz, len);
-      this->apply_timezone_();
+    if (tz == nullptr) {
+      this->apply_timezone_(nullptr);
+      return;
     }
+    // Stack buffer - TZ strings are typically short but allow up to 128
+    char buf[128];
+    if (len >= sizeof(buf))
+      len = sizeof(buf) - 1;
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/time/real_time_clock.h:39]: these come from aioesphomeapi. never happens in practice

File: esphome/components/time/posix_tz.cpp [hunk 24]
```diff
@@ -0,0 +1,452 @@
...
+    return false;
+  }
+
+  int year = internal::epoch_to_year(utc_epoch);
+
+  // Calculate DST start and end for this year
+  // DST start transition happens in standard time
+  time_t dst_start = internal::calculate_dst_transition(year, tz.dst_start, tz.std_offset_seconds);
+  // DST end transition happens in daylight time
+  time_t dst_end = internal::calculate_dst_transition(year, tz.dst_end, tz.dst_offset_seconds);
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/time/posix_tz.cpp:357]: never happens

File: esphome/components/time/posix_tz.cpp [hunk 25]
```diff
@@ -0,0 +1,452 @@
...
+  if (!internal::skip_tz_name(p)) {
+    return false;
+  }
+
+  // Parse standard offset (required)
+  if (!*p || (!std::isdigit(static_cast<unsigned char>(*p)) && *p != '+' && *p != '-')) {
+    return false;
+  }
+  result.std_offset_seconds = internal::parse_offset(p);
+
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/time/posix_tz.cpp:397]: never happens

File: esphome/components/time/real_time_clock.h [hunk 26]
```diff
@@ -20,26 +23,31 @@ class RealTimeClock : public PollingComponent {
...
+  /// Set the time zone from a POSIX TZ string.
+  void set_timezone(const char *tz) { this->apply_timezone_(tz); }
 
-  /// Set the time zone from raw buffer, only if it differs from the current one.
+  /// Set the time zone from a character buffer with known length.
+  /// The buffer does not need to be null-terminated.
   void set_timezone(const char *tz, size_t len) {
-    if (this->timezone_.length() != len || memcmp(this->timezone_.c_str(), tz, len) != 0) {
-      this->timezone_.assign(tz, len);
-      this->apply_timezone_();
```
Comment (swoboda1337) [inline esphome/components/time/real_time_clock.h:33]: Deduplication check removed, not sure how often this gets called.

File: esphome/components/time/posix_tz.cpp [hunk 27]
```diff
@@ -0,0 +1,488 @@
...
+static inline int days_in_year(int year) { return is_leap_year(year) ? 366 : 365; }
+
+// Convert days since epoch to year, updating days to remainder
+static int __attribute__((noinline)) days_to_year(int64_t &days) {
+  int year = 1970;
+  int diy;
+  while (days >= (diy = days_in_year(year))) {
+    days -= diy;
+    year++;
+  }
```
Comment (swoboda1337) [inline esphome/components/time/posix_tz.cpp:43]: If a garbage value was passed in this loop could spin for 5 hours

File: esphome/components/time/real_time_clock.h [hunk 28]
```diff
@@ -20,26 +23,31 @@ class RealTimeClock : public PollingComponent {
...
+    char buf[128];
+    if (len >= sizeof(buf))
+      len = sizeof(buf) - 1;
+    memcpy(buf, tz, len);
+    buf[len] = '\0';
+    this->apply_timezone_(buf);
   }
 
-  /// Get the time zone currently in use.
-  std::string get_timezone() { return this->timezone_; }
```
Comment (swoboda1337) [inline esphome/components/time/real_time_clock.h:38]: This is a breaking change do we need to remove it?

Comment (bdraco) [reply esphome/components/time/real_time_clock.h:38]: Nothing calls this in the entire codebase and its not documented on https://esphome.io/components/time/ so I think its ok to remove

File: esphome/components/time/real_time_clock.h (hunk 26 above)
Comment (bdraco) [reply esphome/components/time/real_time_clock.h:33]: I think this is worth keeping

Comment (bdraco) [reply esphome/components/time/real_time_clock.h:33]: Added deduplication back inside `apply_timezone_()` so all `set_timezone` overloads benefit. It now compares the newly parsed `ParsedTimezone` against the current global via `memcmp` and skips the update if unchanged.

File: esphome/components/time/posix_tz.cpp (hunk 27 above)
Comment (bdraco) [reply esphome/components/time/posix_tz.cpp:43]: Added year bounds (1900–2200) to both loops in `days_to_year()`. This covers any reasonable `time_t` value and prevents runaway iteration from garbage input.

File: esphome/components/time/real_time_clock.h (hunk 28 above)
Comment (bdraco) [reply esphome/components/time/real_time_clock.h:38]: Nothing calls this in the codebase and it's not documented on esphome.io/components/time, so it should be safe to remove.

Comment (swoboda1337) [reply esphome/components/time/real_time_clock.h:38]: People call it lambds, didnt take long to find an example :) 

https://github.com/nathanedudley/Waveshare-ESP32-P4-86-Panel-ETH-2RO_Thermostat/blob/c3465aba38a10d27941059f11e7f3bd90c9e672d/src/pages/home/home.yaml#L2480

File: esphome/components/time/real_time_clock.h (hunk 26 above)
Comment (bdraco) [reply esphome/components/time/real_time_clock.h:33]: But thats kinda pointless because we have to parse it anyways.  And the assign is mostly free since its a POD type.  Removed the `memcmp` check.   The original check was there to avoid the std::string overhead which is gone now anyways

File: esphome/components/time/real_time_clock.h (hunk 28 above)
Comment (bdraco) [reply esphome/components/time/real_time_clock.h:38]: That looks like it was just used for testing.  Its not considered a public API per our published policy https://developers.esphome.io/contributing/code/#what-is-considered-public-c-api. https://developers.esphome.io/contributing/code/#c-user-expectations

Comment (bdraco) [reply esphome/components/time/real_time_clock.h:38]: If we really want to keep it we can keep storing it at python codegen time as a string and not deprecate sending the string from Home Assistant.  However thats RAM + heap churn + ~1000-2000 bytes of flash bloat we would keep around for that forever for something that is likely unused by 99.9% of users.  My preference is to remove it.

Comment (edwardtfn) [reply esphome/components/time/real_time_clock.h:38]: I’m using that in my project, but it’s just cosmetic, so I can easily deal with that.

https://github.com/Blackymas/NSPanel_HA_Blueprint/blob/1dd62617d2213bc77451f5a57aa5d3090075607c/esphome/nspanel_esphome_datetime.yaml#L122

In fact the only reason I deal with time zone is due to the fact time zone is confusing from  ESPHome side and hopefully this PR will help with this.
In the NSPanel we display time, so timezone is more critical than lots of other applications where it is only used internally.

Comment (bdraco) [reply esphome/components/time/real_time_clock.h:38]: Yeah I think the only practical reason someone would call it is for debugging/logging. I don't think anyone is going to realistically parse a posix timezone string in a lambda.

Comment (bdraco) [reply esphome/components/time/real_time_clock.h:38]: The more important long term question is if we need to keep the `set_timezone` char overload and the parser bloat in the binary forever. The plan was to remove all that bloat in https://github.com/esphome/esphome/pull/14237 as well

Comment (edwardtfn) [reply esphome/components/time/real_time_clock.h:38]: In fact that whole timezone management from our side is probably not need anymore since it was changed to get timezone from HA

Comment (bdraco) [reply esphome/components/time/real_time_clock.h:38]: Good to know. It seems we need a way to better communicate changes that affect lambads using undocumented functions

Comment (bdraco) [reply esphome/components/time/real_time_clock.h:38]: Updated this pr to mark it as an `undocumented-api-change`. @edwardtfn If you find you end up needing APIs in the long term to set the timezone externally please let me know on discord and we can work something out.

**Review by swoboda1337 — APPROVED** [review]

**Comment by bdraco** [comment]
Thanks
